pkg/sinks/forward/forward.go

```diff
@@ -222,60 +219,32 @@ func (df *DataForward) forwardAChunk(ctx context.Context) {
 
-	var messageToStep []isb.Message
-
-	// concurrent processing request channel
-	processingCh := make(chan *readWriteMessagePair)
-	// sinkResults stores the results after processing for all read messages. It indexes
-	// a read message to the write message
-	sinkResults := make([]readWriteMessagePair, len(dataMessages))
-
-	// create a pool of processors
-	var wg sync.WaitGroup
-	for i := 0; i < df.opts.sinkConcurrency; i++ {
-		wg.Add(1)
-		go func() {
-			defer wg.Done()
-			df.concurrentConvertReadToWriteMsgs(ctx, processingCh)
-		}()
-	}
-	concurrentProcessingStart := time.Now()
-
-	// only send the data messages
-	for idx, m := range dataMessages {
-		// emit message size metric
-		metrics.ReadBytesCount.With(map[string]string{metrics.LabelVertex: df.vertexName, metrics.LabelPipeline: df.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeSink), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(df.vertexReplica)), metrics.LabelPartitionName: df.fromBufferPartition.GetName()}).Add(float64(len(m.Payload)))
-		// assign watermark to the message
+	writeMessages := make([]isb.Message, 0, len(dataMessages))
+	for _, m := range dataMessages {
 		m.Watermark = time.Time(processorWM)
-		// send message to the channel
-		sinkResults[idx].readMessage = m
-		processingCh <- &sinkResults[idx]
-	}
-	// let the go routines know that there is no more work
-	close(processingCh)
-	// wait till the processing is done. this will not be an infinite wait because the processing will exit if
-	// context.Done() is closed.
-	wg.Wait()
-	df.opts.logger.Debugw("concurrent convert read message to write message completed", zap.Int("concurrency", df.opts.sinkConcurrency), zap.Duration("took", time.Since(concurrentProcessingStart)))
-
-	// let's figure out which vertex to send the results to.
-	// update the toBuffer(s) with writeMessages.
-	for _, m := range sinkResults {
-		// update toBuffer
-		for _, message := range m.writeMessages {
-			messageToStep = append(messageToStep, message.Message)
-		}
+		writeMessages = append(writeMessages, m.Message)
 	}
 
-	// forward the message to the edge buffer (could be multiple edges)
-	writeOffsets, err := df.writeToBuffer(ctx, df.toBuffer, messageToStep)
+	// write the messages to the sink
+	writeOffsets, fallbackMessages, err := df.writeToSink(ctx, df.sinkWriter, writeMessages, false)
+	// error will not be nil only when we get ctx.Done()
 	if err != nil {
-		df.opts.logger.Errorw("failed to write to toBuffer", zap.Error(err))
+		df.opts.logger.Errorw("failed to write to sink", zap.Error(err))
 		df.fromBufferPartition.NoAck(ctx, readOffsets)
 		return
 	}
-	df.opts.logger.Debugw("writeToBuffers completed")
 
+	// if we have any messages to write to the fallback sink, write them.
+	if len(fallbackMessages) > 0 {
+		df.opts.logger.Infow("Writing messages to fallback sink", zap.Int("count", len(fallbackMessages)))
+		_, fallbackMessages, err = df.writeToSink(ctx, df.opts.fbSinkWriter, fallbackMessages, true)
```

Comment:
```suggestion
		// write to sink is an infinite loop; it will return only if writes are successful or
		// ctx.Done happens due to shutdown.
		_, _, err = df.writeToSink(ctx, df.opts.fbSinkWriter, fallbackMessages, true)
```